core/src/main/scala/dagr/core/tasksystem/Retry.scala

```diff
@@ -0,0 +1,137 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package dagr.core.tasksystem
+
+import java.nio.file.Files
+
+import dagr.commons.util.LazyLogging
+import dagr.core.execsystem._
+import dagr.commons.io.Io
+
+/** A trait to facilitate retry a task when it has failed. */
+trait Retry {
+  /** This method will be invoked to allow the task to perform any self-modification before retrying.
+    *
+    * @param resources the maximum system resources available to the task
+    * @param taskInfo the task execution information for the task to be retried.
+    * @return true if the task is to be retried, false otherwise.
+    */
+  def retry(resources: SystemResources, taskInfo: TaskExecutionInfo): Boolean = false
+}
+
+/**
+  * A trait to facilitate retrying a task with more memory.
+  */
+trait MemoryRetry extends Retry with FixedResources {
+  /** Given the current memory, returns the next memory to retry this task with, or None, if the task
+    * should not be retried.
+    */
+  protected def nextMemory(currentMemory: Memory): Option[Memory]
+
+  /** True if we should use the system memory as the maximum memory, false to use the heap size. */
+  protected def useSystemMemory: Boolean = true
+
+  /** Determines if this task ran out of memory, and if so, it will get the next memory value with which to be retried. */
+  protected def ranOutOfMemory(taskInfo: TaskExecutionInfo): Boolean = true
+
+  /** Tries to get the next memory value with which to run task.  It will enforce tha the task will not require more
+    * memory than in the system.
+    * @param systemResources the system resources under which this task should run.
+    * @param taskInfo the task execution information for the task to be retried.
+    * @return true if the task is to be retried, false otherwise.
+    */
+  override final def retry(systemResources: SystemResources, taskInfo: TaskExecutionInfo): Boolean = {
+    if (taskInfo.resources.memory != this.resources.memory) throw new IllegalStateException("Scheduled memory does not equal current memory")
+    if (ranOutOfMemory(taskInfo)) {
+      val maximumMemory = if (useSystemMemory) systemResources.systemMemory else systemResources.jvmMemory
+      nextMemory(this.resources.memory)
+        .filter { memory => memory <= maximumMemory }
```

Comment:
I think this would be slightly clearer as just ` _ <= maximumMemory`, since my mind wants to see the `=>` and `<=` as symmetrical operators.